tests/end_to_end_tests/run_in_remote_machine/ec2.py

```diff
@@ -0,0 +1,361 @@
+import dataclasses
+import logging
+import os
+import pathlib as pl
+import shlex
+import time
+import random
+import json
+from typing import List, Dict, Optional
+
+"""
+This module defines main logic that is responsible for manipulating with ec2 instances to run end to end tests
+inside them.
+"""
+
+logger = logging.getLogger(__name__)
+
+# All the instances created by this script will use this string in the name.
+INSTANCE_NAME_STRING = "automated-agent-tests-"
+assert "-tests-" in INSTANCE_NAME_STRING
+
+MAX_PREFIX_LIST_UPDATE_ATTEMPTS = 20
+
+
+@dataclasses.dataclass
+class EC2DistroImage:
+    """
+    Simple specification of the ec2 AMI image.
+    """
+    image_id: str
+    image_name: str
+    size_id: str
+    ssh_username: str
+
+
+def destroy_node_and_cleanup(driver, node):
+    """
+    Destroy the provided node and cleanup any left over EBS volumes.
+    """
+    volumes = driver.list_volumes(node=node)
+
+    assert (
+            INSTANCE_NAME_STRING in node.name
+    ), "Refusing to delete node without %s in the name" % (INSTANCE_NAME_STRING)
+
+    print("")
+    print(('Destroying node "%s"...' % (node.name)))
+
+    try:
+        node.destroy()
+    except Exception as e:
+        if "does not exist" in str(e):
+            # Node already deleted, likely by another concurrent run. This error is not fatal so we
+            # just ignore it.
+            print(
+                "Failed to delete node, likely node was already deleted, ignoring error..."
+            )
+            print(str(e))
+        else:
+            raise e
+
+    assert len(volumes) <= 1
+    print("Cleaning up any left-over EBS volumes for this node...")
+
+    # Give it some time for the volume to become detached from the node
+    if volumes:
```

Comment:
I believe there is a way to check volume status, but it can be a TODO

Reply:
Added volumes status check.